### PR TITLE
2.x

### PR DESCRIPTION
**Pull request by atz**

### CODE REVIEW

**Comment by atz** [comment]
@darrenleeweber

File: README.md [hunk 1]
```diff
@@ -8,18 +8,43 @@ https://consul.stanford.edu/display/DOR/DOR+services#DORservices-initializeworkf
...
+  faraday.adapter  Faraday.default_adapter  # make requests with Net::HTTP
+end
+wfs = Dor::WorkflowService.new(conn)
+```
+
+The corresponding client will be constructed off the same URL.  If this is insufficient, you can always set a client directly:
+
+```ruby
+wfs.resource = RestClient::Resource('http://protected/resource', :user => 'user', :password => 'password', :read_timeout => 10)
+```
```
Comment (darrenleeweber) [inline README.md]: The use of two HTTP clients is confusing and likely a maintenance headache.  What's the rationale for this?  Can this gem use one or the other?  The general consensus in DLSS has been to prefer faraday over rest-client.

File: lib/dor/services/workflow_service.rb [hunk 2]
```diff
@@ -7,539 +7,450 @@
...
+    # Initialize the workflow service
+    # @param [String, Faraday, RestClient::Resource] url or configured Faraday/RestClient::Resource object that points to the workflow service
+    # @param [Hash] opts optional params
+    # @option opts [String]  :dor_services_url uri to the DOR REST service
+    # @option opts [Logger]  :logger defaults writing to workflow_service.log with weekly rotation
+    # @option opts [Integer] :timeout number of seconds for Faraday timeout, default: 180
+    # @option opts [Proc]    :handler code triggered by Faraday exception, receives the exception object, attempt_number and total_delay.  See retries gem.
+    # @option opts [Array<Class>] :exceptions_to_catch classes of exceptions to be caught by retry handler
+    #
+    # TODO: convert RestClient use to Faraday
```
Comment (darrenleeweber) [inline lib/dor/services/workflow_service.rb]: So do it now?

File: lib/dor/services/workflow_service.rb [hunk 3]
```diff
@@ -7,539 +7,450 @@
...
+      @handler             = opts[:handler] || proc do |exception, attempt_number, total_delay|
+        @logger.warn "[Attempt #{attempt_number}] #{exception.class}: #{exception.message}; #{total_delay} seconds elapsed."
+      end
+      url_string = url.is_a?(RestClient::Resource) ? url.url :
+                   url.is_a?(Faraday) ? url.url_prefix.to_s : url
+      @resource = url.is_a?(RestClient::Resource) ? url : RestClient::Resource.new(url_string, params)
+      @http_conn = url.is_a?(Faraday) ? url : Faraday.new(url: url_string) do |faraday|
+        faraday.response :logger, @logger if opts[:debug]
+        faraday.adapter  :net_http_persistent    # use Keep-Alive connections
       end
```
Comment (darrenleeweber) [inline lib/dor/services/workflow_service.rb]: Make the choice to use faraday and drop rest-client in the init.

File: lib/dor/services/workflow_service.rb [hunk 4]
```diff
@@ -7,539 +7,450 @@
...
+    # @param [String] druid The id of the object
+    # @param [String] workflow The name of the workflow
+    # @param [String] process The name of the process step
+    # @return [String] status for repo-workflow-process-druid
+    def get_workflow_status(repo, druid, workflow, process)
+      doc = get_workflow_ngxml(repo, druid, workflow)
+      status = doc.root.at_xpath("//process[@name='#{process}']/@status")
+      status = status.content if status
+      status
+    end
```
Comment (darrenleeweber) [inline lib/dor/services/workflow_service.rb]: This is hard to compare with the prior method, only because it has moved location in the file.  In general, unless there are compelling reasons to group similar method, an alphabetical order of methods is easier to follow.  The exception is often to group the get/set methods that share the same suffix, otherwise the methods are ordered alphabetically by the suffix.  What is the significance of the `update` prefix vs. a `set` prefix to match the `get` prefixes?  I tend to equate a PUT with `set` and a PATCH with `update`.

File: lib/dor/services/workflow_service.rb [hunk 5]
```diff
@@ -7,539 +7,450 @@
...
+    # @option opts [Float] :elapsed The number of seconds it took to complete this step. Can have a decimal.  Is set to 0 if not passed in.
+    # @option opts [String] :lifecycle Bookeeping label for this particular workflow step.  Examples are: 'registered', 'shelved'
+    # @option opts [String] :note Any kind of string annotation that you want to attach to the workflow
+    # @option opts [String] :lane_id Id of processing lane used by the job manager.  Can convey priority or name of an applicaiton specific processing lane (e.g. 'high', 'critical', 'hydrus')
+    # @option opts [String] :current_status Setting this string tells the workflow service to compare the current status to this value.  If the current value does not match this value, the update is not performed
+    # @return []
+    def update_workflow_status(repo, druid, workflow, process, status, opts = {})
+      raise ArgumentError, "Unknown status value #{status}" unless @valid_statuses.include?(status.downcase)
+      opts = { :elapsed => 0, :lifecycle => nil, :note => nil }.merge!(opts)
+      opts[:elapsed] = opts[:elapsed].to_s
```
Comment (darrenleeweber) [inline lib/dor/services/workflow_service.rb]: `:elapsed` is init as integer.  The merge will take whatever the parameter provides.  Curious that it's cooerced to a string afterwards and not an integer (as in the init).

File: lib/dor/services/workflow_service.rb [hunk 6]
```diff
@@ -7,539 +7,450 @@
...
+      current_status = opts.delete(:current_status)
```
Comment (darrenleeweber) [inline lib/dor/services/workflow_service.rb]: current_status could be `nil` if it's not in opts.

File: lib/dor/services/workflow_service.rb [hunk 7]
```diff
@@ -7,539 +7,450 @@
...
+      xml = create_process_xml({ :name => process, :status => status.downcase }.merge!(opts))
+      uri = "#{repo}/objects/#{druid}/workflows/#{workflow}/#{process}"
+      uri << "?current-status=#{current_status.downcase}" if current_status
+      workflow_resource_method(uri, 'put', xml, { :content_type => 'application/xml' })
+    end
```
Comment (darrenleeweber) [inline lib/dor/services/workflow_service.rb]: `current_status` could be `nil` if it's not in opts, so the `uri` parameter could be missing and have to assume that it's optional for this REST request.

File: lib/dor/services/workflow_service.rb [hunk 8]
```diff
@@ -7,539 +7,450 @@
...
+    # @param [String] process The name of the process step
+    # @param [String] status The status that you want to set -- using one of the values in @valid_statuses
+    # @param [Hash] opts optional values for the workflow step
+    # @option opts [Float] :elapsed The number of seconds it took to complete this step. Can have a decimal.  Is set to 0 if not passed in.
+    # @option opts [String] :lifecycle Bookeeping label for this particular workflow step.  Examples are: 'registered', 'shelved'
+    # @option opts [String] :note Any kind of string annotation that you want to attach to the workflow
+    # @option opts [String] :lane_id Id of processing lane used by the job manager.  Can convey priority or name of an applicaiton specific processing lane (e.g. 'high', 'critical', 'hydrus')
+    # @option opts [String] :current_status Setting this string tells the workflow service to compare the current status to this value.  If the current value does not match this value, the update is not performed
+    # @return []
+    def update_workflow_status(repo, druid, workflow, process, status, opts = {})
```
Comment (darrenleeweber) [inline lib/dor/services/workflow_service.rb]: Wrap the help documentation lines or leave it to yard?

File: lib/dor/services/workflow_service.rb (hunk 7 above)
Comment (darrenleeweber) [inline lib/dor/services/workflow_service.rb]: If all the workflow HTTP requests use `{ :content_type => 'application/xml' }`, then set it in the HTTP client init rather than every method.

File: spec/workflow_service_spec.rb [hunk 9]
```diff
@@ -27,56 +27,48 @@
...
-
+    allow(@mock_logger).to receive(:info)
+    allow(@mock_logger).to receive(:warn)
+    allow(@mock_logger).to receive(:debug) # silence logging output
+    @mock_http_connection = double('mock_http_connection')
     allow(Faraday).to receive(:new).and_return(@mock_http_connection)
-
-    allow(RestClient::Resource).to receive(:new).and_return(@mock_resource)
-    Dor::WorkflowService.configure @mock_resource.url
+    @dwfs = Dor::WorkflowService.new(@mock_resource, :logger => @mock_logger)
```
Comment (darrenleeweber) [inline spec/workflow_service_spec.rb]: Again, the use of two HTTP clients is maintenance overhead that's not necessary.  Use faraday, drop rest-client.

File: spec/workflow_service_spec.rb [hunk 10]
```diff
@@ -27,56 +27,48 @@
...
       expect(@mock_resource).to receive(:put).twice {
         runs += 1
         raise ex if runs == 1
         raise Exception.new('Something Else Happened') if runs == 2
       }
-      expect(@mock_logger).to receive(:warn).with(/\[Attempt 1\] RestClient::Exception: #{ex.message}/)
-      expect{ Dor::WorkflowService.create_workflow(@repo, @druid, 'etdSubmitWF', wf_xml) }.to raise_error(Exception, 'Something Else Happened')
+      expect(@mock_logger).to receive(:warn).with(/\[Attempt 1\] Faraday::ClientError/)
+      expect{ @dwfs.create_workflow(@repo, @druid, 'etdSubmitWF', wf_xml) }.to raise_error(Exception, 'Something Else Happened')
     end
```
Comment (darrenleeweber) [inline spec/workflow_service_spec.rb]: Set the retry parameter on init for Faraday and set the logger on Faraday and leave all of this up to it, no need to spec basic HTTP client behavior here.

File: spec/workflow_service_spec.rb (hunk 9 above)
Comment (darrenleeweber) [inline spec/workflow_service_spec.rb]: Most of the changes in the specs are a global replace of `Dor::WorkflowService` for `@dwfs`.  Consider using `let` instead of instance vars and whether the `@dwfs` could be `subject`.

**Comment by darrenleeweber** [comment]
:0: on the use of two HTTP clients.  The change from a module to a class is OK and that's the core of this PR, with implications for anything using the gem.  I'll have to keep an eye on this PR and evaluate implications for SDR-PC robots.

**Comment by drh-stanford** [comment]
what's the status on this PR? can you summarize its changes as a major new release (2.x) -- is it only the new configuration API?

File: lib/dor/services/workflow_service.rb (hunk 8 above)
Comment (atz) [inline lib/dor/services/workflow_service.rb]: let Yard do it

File: lib/dor/services/workflow_service.rb (hunk 5 above)
Comment (atz) [inline lib/dor/services/workflow_service.rb]: It's because it is going into XML as a string.

**Comment by atz** [comment]
I think @cbeer is cherry-picking the meat of this PR to other branches.  This may end up targeting 3.x, even though it started the 2.x branch.

**Comment by cbeer** [comment]
Yes, these changes are incorporated into #30.
